src/main/java/com/commercetools/sync/commons/utils/QueryAll.java

```diff
@@ -1,166 +1,180 @@
+        if (currentPageStage != null) {
+            return currentPageStage
+                .thenCompose(currentPage -> queryNextPages(processPageAndGetNext(currentPage)));
+        }
+        return completedFuture(this);
     }
 
     /**
-     * Given a callback {@link Function}, this method first calculates the total number of pages resulting from the
-     * query, then it applies this callback on each page of the results from {@link this} instance's {@code baseQuery}
-     * and returns a list containing the result of this callback on each page.
+     * Given a result {@link PagedQueryResult}, this method checks if there are elements in the result (size > 0), then
+     * it maps or consumes the resultant list using this instance's {@code pageMapper} or {code pageConsumer} whichever
+     * is available. Then, it checks if this page is the last page or not (by checking if the result size is equal to
+     * this instance's {@code pageSize}). If It is, then it means there might be still more results. However, if
+     * it is less, then it means for sure there are no more results and this is the last page. If there is a next page,
+     * then a new future of the next page is returned. If there are no more results, the method returns null.
      *
-     * @param client        the CTP client that the query is run on.
-     * @param totalElements the total number of elements resulting from the query.
-     * @param callback      the callback to apply on each page of results.
-     * @return a list of futures containing the results of applying the callback on each page of results.
+     * @param page the current page result.
+     * @return If there is a next page, then a new future of the next page is returned. If there are no more results,
+     *         the method returns null.
      */
-    @Nonnull
-    private <S> List<CompletableFuture<S>> queryNextPages(final SphereClient client, final long totalElements,
-                                                          final Function<List<T>, S> callback) {
-        final long totalPages = getTotalNumberOfPages(totalElements);
-        return LongStream.range(1, totalPages)
-                         .mapToObj(page -> queryPage(client, page)
-                             .thenApply(PagedQueryResult::getResults)
-                             .thenApply(callback)
-                             .toCompletableFuture())
-                         .collect(toList());
+    @Nullable
+    private CompletionStage<PagedQueryResult<T>> processPageAndGetNext(@Nonnull final PagedQueryResult<T> page) {
+        final List<T> currentPageElements = page.getResults();
+        if (currentPageElements.size() == pageSize) {
+            mapOrConsume(currentPageElements);
+            return getNextPageStage(currentPageElements);
+        } else {
+            if (currentPageElements.size() > 0) {
+                mapOrConsume(currentPageElements);
+            }
+        }
+        return null;
```

Comment:
could be simplified like this?
```java
if (currentPageElements.size() > 0) {
    mapOrConsume(currentPageElements);
}

return getNextPageStage(currentPageElements);
```

and then 
```java
private CompletionStage<PagedQueryResult<T>> getNextPageStage(@Nonnull final List<T> pageElements) {
    if (pageElements.size() == pageSize) {
        final String lastElementId = pageElements.get(pageElements.size() - 1).getId();
        final QueryPredicate<T> queryPredicate = QueryPredicate.of(format("id > \"%s\"", lastElementId));
        return queryPage(queryPredicate);
    }

    return completedFuture(null);
}
```

IMHO (but not mandatory) this looks a bit more separated: `processPageAndGetNext` cares about processing and getting, `getNextPageStage` cares about page result

Reply:
2638c02